agent/consul/gateways/controller_gateways_test.go

```diff
@@ -0,0 +1,123 @@
+package gateways
+
+import (
+	"context"
+	"github.com/hashicorp/consul/acl"
+	"github.com/hashicorp/consul/agent/consul/controller"
+	"github.com/hashicorp/consul/agent/consul/gateways/datastore"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/go-hclog"
+	"github.com/stretchr/testify/mock"
+	"testing"
+)
+
+func Test_apiGatewayReconciler_Reconcile(t *testing.T) {
+	type fields struct {
+		logger hclog.Logger
+		store  datastore.DataStore
+	}
+	type args struct {
+		ctx context.Context
+		req controller.Request
+	}
+	tests := []struct {
+		name    string
+		fields  fields
+		args    args
+		wantErr bool
+	}{
+		{
+			name: "happy path - update available",
+			fields: fields{
+				store:  datastoreWithUpdate(t),
+				logger: hclog.Default(),
+			},
+			args: args{
+				ctx: context.Background(),
+				req: controller.Request{
+					Kind: structs.APIGateway,
+					Name: "test-gateway",
+					Meta: acl.DefaultEnterpriseMeta(),
+				},
+			},
+			wantErr: false,
+		},
+		{
+			name: "delete happy path",
+			fields: fields{
+				store:  datastoreWithDelete(t),
+				logger: hclog.Default(),
+			},
+			args: args{
+				ctx: context.Background(),
+				req: controller.Request{
+					Kind: structs.APIGateway,
+					Name: "test-gateway",
+					Meta: acl.DefaultEnterpriseMeta(),
+				},
+			},
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			r := apiGatewayReconciler{
+				logger: tt.fields.logger,
+				store:  tt.fields.store,
+			}
+			if err := r.Reconcile(tt.args.ctx, tt.args.req); (err != nil) != tt.wantErr {
```

Comment:
are we planning on adding more tests around the sad path? if not can we simplify the error check to just  be `if err != nil`, if not would it make sense to use any sort of sentinel errors and use `errors.Is` around the type?

Reply:
Definitely planning on adding more tests in the future

Reply:
Probably would make sense, pretty sure this was just the autogenerated boilerplate test structure